core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala

```diff
@@ -77,5 +80,124 @@ class BrokerMetadataListenerTest {
       listener.close()
     }
   }
+
+  class MockMetadataSnapshotter extends MetadataSnapshotter {
+    var image = MetadataImage.EMPTY
+    val failure = new AtomicReference[Throwable](null)
+    var activeSnapshotOffset = -1L
+    var prevCommittedOffset = -1L
+    var prevCommittedEpoch = -1
+    var prevLastContainedLogTime = -1L
+
+    override def maybeStartSnapshot(committedOffset: Long,
+                                    committedEpoch: Int,
+                                    lastContainedLogTime: Long,
+                                    newImage: MetadataImage): Boolean = {
+      try {
+        if (activeSnapshotOffset == -1L) {
+          assertTrue(prevCommittedOffset <= committedOffset)
+          assertTrue(prevCommittedEpoch <= committedEpoch)
+          assertTrue(prevLastContainedLogTime <= lastContainedLogTime)
+          prevCommittedOffset = committedOffset
+          prevCommittedEpoch = committedEpoch
+          prevLastContainedLogTime = lastContainedLogTime
+          image = newImage
+          activeSnapshotOffset = committedOffset
+          true
+        } else {
+          false
+        }
+      } catch {
+        case t: Throwable => failure.compareAndSet(null, t)
+      }
+    }
+  }
+
+  class MockMetadataPublisher extends MetadataPublisher {
+    var image = MetadataImage.EMPTY
+
+    override def publish(newHighestMetadataOffset: Long,
+                         delta: MetadataDelta,
+                         newImage: MetadataImage): Unit = {
+      image = newImage
+    }
+  }
+
+  private val FOO_ID = Uuid.fromString("jj1G9utnTuCegi_gpnRgYw")
+
+  private def generateManyRecords(listener: BrokerMetadataListener,
+                                  endOffset: Long): Unit = {
+    (0 to 10000).foreach { _ =>
+      listener.handleCommit(RecordTestUtils.mockBatchReader(endOffset,
```

Comment:
Not sure it matters, but it is a little strange that we use the same end offset for all of these records.

Reply:
It makes it easier since otherwise I'd have to figure out exactly which record triggered the snapshot... that would make the test brittle as well since record sizes will change as fields are added, etc.